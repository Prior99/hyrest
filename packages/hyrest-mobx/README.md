# Hyrest MobX

<img align="right" width="200" height="200" src="https://github.com/Prior99/hyrest/raw/master/logo/hyrest-logo-400px.png">

[![npm](https://img.shields.io/npm/v/hyrest-mobx.svg)](https://www.npmjs.com/package/hyrest-mobx)
[![pipeline status](https://gitlab.com/prior99/hyrest/badges/master/pipeline.svg)](https://github.com/Prior99/hyrest)
[![coverage report](https://gitlab.com/prior99/hyrest/badges/master/coverage.svg)](https://github.com/Prior99/hyrest)

Hyrest is a hybrid REST framework for both the client and the server.

**This is the MobX frontend utility package.**

After creating and exposing a REST Api using [hyrest](../hyrest) and [hyrest-express](../hyrest-express) it might be useful to re-use the existing models and validation from the server side in the client to store data next to forms and perform validation on it.

## Usage

Let's say a model has been defined which is already used by a controller:

```typescript
import { is, email, length } from "hyrest";
class User {
    @is().validate(email) public email: string;
    @is().validate(length(5, 20)) public name: string;
    @is().validate(length({ min: 10 })) public password: string;
}
```

In order to create a new user, one might want to implement a sign up form. This is where this package can come in handy.

Instead of creating a `@observable public email: string` and so on for each field on the `User` model, simple utilize hyrest-mobx:

```typescript
import * as React from "react";
import { observer } from "mobx-react";
import { field, hasFields, Field } from "hyrest-mobx";

@observer @hasFields()
class Signup extends React.Component {
    @field(User) private user: Field<User>;

    @action.bound private handleSubmit() {
        console.log(user.value); // Will log the unwrapped user.
    }

    public render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <input type="text" {...this.user.nested.email.reactInput} />
                <input type="text" {...this.user.nested.name.reactInput} />
                <input type="password" {...this.user.nested.password.reactInput} />
                {
                    this.users.errors.map((err, index) => <p key={index}>{err}</p>)
                }
                <button disabled={!this.user.valid}>Signup</button>
            </form>
        );
    }
}
```

The decorator `@field` will mark the property as a field with the type `User`.

After marking the class with `@hasFields`, new `Field`s will automatically be injected upon creation.

The injected values already utilize Mobx under the hood and hence no `@observable`s are no longer necessary.

## Resources

- [Tutorial](https://prior99.gitlab.io/hyrest/docs/tutorial-about)
- [Minimal example project](https://github.com/Prior99/hyrest-todo-example)
- [Documentation](https://prior99.gitlab.io/hyrest/)
- [Guide](https://prior99.gitlab.io/hyrest/docs/preamble-about/)
- [API Reference](https://prior99.gitlab.io/hyrest/api/hyrest-mobx/)
