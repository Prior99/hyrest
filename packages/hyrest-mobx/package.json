{
  "name": "hyrest-mobx",
  "version": "0.0.0",
  "description": "A hybrid REST library for the frontend and backend. Mobx connector package.",
  "keywords": [
    "hyrest",
    "hybrid",
    "http",
    "rest",
    "api",
    "promise",
    "form",
    "field",
    "mobx"
  ],
  "main": "dist/index.js",
  "author": "Frederick Gnodtke",
  "license": "MIT",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "docs": "typedoc --mode file --out docs/ src/"
  },
  "typings": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/Prior99/hyrest"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md"
  ],
  "devDependencies": {
    "typedoc": "^0.11.1",
    "typescript": "^3.0.1"
  },
  "dependencies": {
    "bind-decorator": "^1.0.11",
    "hyrest": "^0.8.0",
    "mobx": "^5.0.3",
    "reflect-metadata": "^0.1.12"
  }
}
