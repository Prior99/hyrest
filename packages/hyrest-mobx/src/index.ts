export * from "./field";
export * from "./decorators";
export * from "./types";
export * from "./validation-status";
