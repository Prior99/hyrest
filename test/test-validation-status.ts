import { ValidationStatus, mergeValidationStatus, combineValidationStatus } from "hyrest-mobx";

const allStatus = [
    ValidationStatus.IN_PROGRESS,
    ValidationStatus.VALID,
    ValidationStatus.INVALID,
    ValidationStatus.UNTOUCHED,
];

allStatus.forEach(statusA => {
    allStatus.forEach(statusB => {
        test(`mergeValidationStatus(${statusA}, ${statusB})`, () => {
            expect(mergeValidationStatus(statusA, statusB)).toMatchSnapshot();
        });
    });
});

[
    [ValidationStatus.IN_PROGRESS, ValidationStatus.INVALID, ValidationStatus.INVALID],
    [ValidationStatus.VALID, ValidationStatus.INVALID, ValidationStatus.UNTOUCHED],
    [ValidationStatus.IN_PROGRESS, ValidationStatus.INVALID, ValidationStatus.VALID, ValidationStatus.UNTOUCHED],
    [],
].forEach(statusArray => {
    test(`combineValidationStatus(${statusArray.join(", ")})`, () => {
        expect(combineValidationStatus(statusArray)).toMatchSnapshot();
    });
});
