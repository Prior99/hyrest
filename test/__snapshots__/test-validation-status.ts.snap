// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combineValidationStatus() 1`] = `"untouched"`;

exports[`combineValidationStatus(in progress, invalid, invalid) 1`] = `"in progress"`;

exports[`combineValidationStatus(in progress, invalid, valid, untouched) 1`] = `"in progress"`;

exports[`combineValidationStatus(valid, invalid, untouched) 1`] = `"invalid"`;

exports[`mergeValidationStatus(in progress, in progress) 1`] = `"in progress"`;

exports[`mergeValidationStatus(in progress, invalid) 1`] = `"in progress"`;

exports[`mergeValidationStatus(in progress, untouched) 1`] = `"in progress"`;

exports[`mergeValidationStatus(in progress, valid) 1`] = `"in progress"`;

exports[`mergeValidationStatus(invalid, in progress) 1`] = `"in progress"`;

exports[`mergeValidationStatus(invalid, invalid) 1`] = `"invalid"`;

exports[`mergeValidationStatus(invalid, untouched) 1`] = `"invalid"`;

exports[`mergeValidationStatus(invalid, valid) 1`] = `"invalid"`;

exports[`mergeValidationStatus(untouched, in progress) 1`] = `"in progress"`;

exports[`mergeValidationStatus(untouched, invalid) 1`] = `"invalid"`;

exports[`mergeValidationStatus(untouched, untouched) 1`] = `"untouched"`;

exports[`mergeValidationStatus(untouched, valid) 1`] = `"untouched"`;

exports[`mergeValidationStatus(valid, in progress) 1`] = `"in progress"`;

exports[`mergeValidationStatus(valid, invalid) 1`] = `"invalid"`;

exports[`mergeValidationStatus(valid, untouched) 1`] = `"untouched"`;

exports[`mergeValidationStatus(valid, valid) 1`] = `"valid"`;
